### PR TITLE
apply duct-tape to `InVMKafkaCluster.start` to ensure all nodes are p…

### PR DESCRIPTION
**Pull request by SamBarker**
…resent in the cluster before injecting.

Fixes: #7

### CODE REVIEW

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 1]
```diff
@@ -47,6 +52,7 @@ public class InVMKafkaCluster implements KafkaCluster {
     private final ZooKeeperServer zooServer;
     private final List<Server> servers;
     private final List<String> bootstraps = new ArrayList<>();
+    private final Supplier<KafkaClusterConfig.KafkaEndpoints> kafkaEndpointsSupplier;
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:55]: I'm not entirely convinced this is the right place to keep this. 

It should maybe be stored in the clusterConfig and accessed via there.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 2]
```diff
@@ -109,8 +114,7 @@ public EndpointPair getControllerEndpoint(int brokerId) {
 
             servers = clusterConfig.getBrokerConfigs(kafkaEndpointsSupplier, zookeeperEndpointSupplier).map(this::buildKafkaServer).collect(Collectors.toList());
 
-        }
-        catch (IOException e) {
+        } catch (IOException e) {
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:117]: Must update my formatter settings to match the maven plug-in (thought I don't like the line break it adds)

File: api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java [hunk 3]
```diff
@@ -60,4 +62,9 @@ public interface KafkaCluster extends AutoCloseable {
      * @return mutable configuration map
      */
     Map<String, Object> getKafkaClientConfiguration(String user, String password);
+
+    /**
+     *
+     */
+    assertEachBrokerDescibesCluster(Predicate<org.apache.kafka.clients.admin.DescribeClusterResult> descibeClusterResultMatcher, Duration timeout);
```
Comment (k-wall) [inline api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java:69]: Is that actually used?

Comment (SamBarker) [reply api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java:69]: Not yet. 

I need to add a test to verify the behaviour.

**Comment by k-wall** [comment]
Do we actually need this PR?  I have a memory that we agreed that it is the role of the test itself to await the cluster to come in coordination. IIRC the argument (expressed by @tombentley) was that the fact Kafka uses an eventually consistent model is inherent.  It would be wrong from a testing framework to hide that from test authors.

**Comment by SamBarker** [comment]
> Do we actually need this PR? I have a memory that we agreed that it is the role of the test itself to await the cluster to come in coordination (https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/7#issuecomment-1337253584) . IIRC the argument (expressed by @tombentley) was that the fact Kafka uses an eventually consistent model is inherent

I think we have largely the same picture of what was agreed. I'm not trying to hide the distributed nature of Kafka however I do think the extension should ensure that the cluster is "ready" for use before handing it over to tests. I.e. if the cluster  is annotated with a broker count of `2` we should ensure `2` brokers are active prior to injection. Which is what I was trying to achieve here. Then provide some supporting code for more sensitive tests that actually want to ensure the coherence of the cluster.

**Comment by k-wall** [comment]
> > Do we actually need this PR? I have a memory that we agreed that it is the role of the test itself to await the cluster to come in coordination ([#7 (comment)](https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/7#issuecomment-1337253584)) . IIRC the argument (expressed by @tombentley) was that the fact Kafka uses an eventually consistent model is inherent
> 
> I think we have largely the same picture of what was agreed. I'm not trying to hide the distributed nature of Kafka however I do think the extension should ensure that the cluster is "ready" for use before handing it over to tests. I.e. if the cluster is annotated with a broker count of `2` we should ensure `2` brokers are active prior to injection. 

That was the view I held too. That's why I had initially add the flag file (being removed by #18)  stuff to the containers so they checked just that.  However, I think @tombentley arguments against that being the default behaviour are sound.  

> Which is what I was trying to achieve here. Then provide some supporting code for more sensitive tests that actually want to ensure the coherence of the cluster.

@tombentley point is that testing for cluster being complete should be **opt-in** at the test level.  Maybe a test annotation would be a unobtrusive way to achieve that? WDYT?

**Comment by franvila** [comment]
> > > Do we actually need this PR? I have a memory that we agreed that it is the role of the test itself to await the cluster to come in coordination ([#7 (comment)](https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/7#issuecomment-1337253584)) . IIRC the argument (expressed by @tombentley) was that the fact Kafka uses an eventually consistent model is inherent
> > 
> > 
> > I think we have largely the same picture of what was agreed. I'm not trying to hide the distributed nature of Kafka however I do think the extension should ensure that the cluster is "ready" for use before handing it over to tests. I.e. if the cluster is annotated with a broker count of `2` we should ensure `2` brokers are active prior to injection.
> 
> That was the view I held too. That's why I had initially add the flag file (being removed by #18) stuff to the containers so they checked just that. However, I think @tombentley arguments against that being the default behaviour are sound.
> 
> > Which is what I was trying to achieve here. Then provide some supporting code for more sensitive tests that actually want to ensure the coherence of the cluster.
> 
> @tombentley's point is that testing for cluster being complete should be **opt-in** at the test level. Maybe a test annotation would be a unobtrusive way to achieve that? WDYT? @franvila?

I agree with @tombentley. We should always avoid to modify the cluster behavior inside the test framework, all the steps needed should be included in each test, as not all kind of tests will need the same steps.

A good test principle to follow is always avoid assertions in the main code, the assertion should always be part of the test itself and it should be unique, that is there should be only one assertion in each test as it shall be the outcome of the test.

**Comment by SamBarker** [comment]
I think by default if the user has requested a `2` broker cluster the extension should ensure there are `2` active brokers *before* it injects them, otherwise every test has to assert the framework has done its job before it can actually test what it is trying to do. Though I'm content to accept that that the framework doesn't need to ensure that the brokers are consistent and fully replicated.

There is possibly an argument for making allowing tests to opt out of that guarantee or a separate mechanism for creating clusters which are not fully functional.

**Comment by SamBarker** [comment]
This PR was based around using `org.rnorth.duct-tape:duct-tape` in order to avoid a dependency on `org.awaitility:org.awaitility` which we now have anyway. I'm going to close this and revisit.

**Pull request closed without merging**